contracts/wrapped_azero/lib.rs

```diff
@@ -0,0 +1,252 @@
+    };
+
+    pub const BALANCE_OF_SELECTOR: [u8; 4] = [0x65, 0x68, 0x38, 0x2f];
+    pub const TRANSFER_SELECTOR: [u8; 4] = [0xdb, 0x20, 0xf9, 0xf5];
+    pub const TRANSFER_FROM_SELECTOR: [u8; 4] = [0x54, 0xb3, 0xc7, 0x6e];
+    pub const ALLOWANCE_SELECTOR: [u8; 4] = [0x4d, 0x47, 0xd9, 0x21];
+
+    #[ink(storage)]
+    #[derive(Default, SpreadAllocate, Storage)]
+    pub struct WrappedAzero {
+        #[storage_field]
+        psp22: psp22::Data,
+        #[storage_field]
+        metadata: metadata::Data,
+        /// access control contract
+        access_control: AccountId,
+    }
+
+    impl PSP22 for WrappedAzero {}
+
+    impl PSP22Metadata for WrappedAzero {}
+
+    // emit events
+    // https://github.com/w3f/PSPs/blob/master/PSPs/psp-22.md
+    impl Internal for WrappedAzero {
+        fn _emit_transfer_event(
+            &self,
+            _from: Option<AccountId>,
+            _to: Option<AccountId>,
+            _amount: Balance,
+        ) {
+            WrappedAzero::emit_event(
+                self.env(),
+                Event::Transfer(Transfer {
+                    from: _from,
+                    to: _to,
+                    value: _amount,
+                }),
+            );
+        }
+
+        fn _emit_approval_event(&self, _owner: AccountId, _spender: AccountId, _amount: Balance) {
+            WrappedAzero::emit_event(
+                self.env(),
+                Event::Approval(Approval {
+                    owner: _owner,
+                    spender: _spender,
+                    value: _amount,
+                }),
+            );
+        }
+    }
+
+    impl AccessControlled for WrappedAzero {
+        type ContractError = PSP22Error;
+    }
+
+    /// Result type
+    pub type Result<T> = core::result::Result<T, PSP22Error>;
+    /// Event type
+    pub type Event = <WrappedAzero as ContractEventBase>::Type;
+
+    /// Event emitted when a token transfer occurs.
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct Transfer {
+        #[ink(topic)]
+        pub from: Option<AccountId>,
+        #[ink(topic)]
+        pub to: Option<AccountId>,
+        pub value: Balance,
+    }
+
+    /// Event emitted when an approval occurs that `spender` is allowed to withdraw
+    /// up to the amount of `value` tokens from `owner`.
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct Approval {
+        #[ink(topic)]
+        owner: AccountId,
+        #[ink(topic)]
+        spender: AccountId,
+        value: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct Wrapped {
+        caller: AccountId,
+        amount: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct UnWrapped {
+        caller: AccountId,
+        amount: Balance,
+    }
+
+    impl WrappedAzero {
+        /// Creates a new token
+        ///
+        /// The token will have its name and symbol set in metadata to the specified values.
+        /// Decimals are fixed at 18.
```

Comment:
```suggestion
        /// Decimals are fixed at 12.
```
as per https://github.com/Cardinal-Cryptography/aleph-node/pull/612/files#diff-40606916260c70af8efd0229d64825968cce2842ab0eee83fa92eeb42646bc24R147